File: spark-event-ui/src/app/model/trackers/sql/SparkPlanNode.ts
import {SparkPlanInfoTree} from '../SparkPlanNode';
import {MetadataKeyValue, MetadataKeyValueMap, SparkPlanInfo} from '../../sparkevents/SparkPlanInfo';
import {MetricValueHolder} from '../metrics/MetricValueHolder';
import {SparkPlanNodeBuilder} from './SparkPlanNodeBuilder';


/**
 * view wrapper for SparkPlanInfo node to add UI specific fields,
 * assign unique IDs to nodes and metricsValueHolders,
 * build a DAG (tree with repetition, if any is detected),
 * compute synthetic metrics, ...
 *
 */
export abstract class SparkPlanNode {
  readonly tree: SparkPlanInfoTree;   // for 1/ id->MetricValueHolder map, and 2/ View UI settings
  readonly id: number;
  readonly parent: SparkPlanNode|undefined;

  // Deprecated (redundant...) use specific subclasses to parse+extract specific fields
  // readonly info: SparkPlanInfo;
  readonly nodeName: string;
  readonly simpleString: string;
  readonly metadata: MetadataKeyValueMap;


  readonly children: SparkPlanNode[];
  readonly metricValueHolders: MetricValueHolder[];

  countMetricValuesToShow(): number { return this.metricValueHolders.filter(x => x.isValueToShow()).length; }

  expanded = true;

  constructor(
    builder: SparkPlanNodeBuilder,
    parent: SparkPlanNode|undefined,
    info: SparkPlanInfo
  ) {
    this.tree = builder.tree;
    this.id = builder.newId();
    this.parent = parent;
    // this.info = info; // Deprecated
    this.nodeName = info.nodeName;
    this.simpleString = info.simpleString;
    this.metadata = info.metadata;

    this.children = (info.children)?
      info.children.map(x => {
        if (x.nodeName === 'InputAdapter' && x.children?.length === 1) {
          // remove intermediate "InputAdapter" node
          return builder.createNode(this, x.children[0]);
        }
        return builder.createNode(this, x);
      }) : [];
    this.metricValueHolders = (info.metrics)? info.metrics.map(x => new MetricValueHolder(this, x, builder.previousTree)) : [];
  }

  getMetadataKeyValues(): MetadataKeyValue[] {
    if (! this.metadata) { return []; }
    // transform javascript Object {} to array of key,value
    return Object.entries(this.metadata).map(([k,v]) => ({ k, v}));
  }

  metricByName(name: string): MetricValueHolder|undefined { return this.metricValueHolders.find(x => x.name === name);}
  metricByAccumulatorId(accumulatorId: number): MetricValueHolder|undefined { return this.metricValueHolders.find(x => x.accumulatorId === accumulatorId);}

  countMetricsWithUpdates(): number {
    return this.metricValueHolders.reduce(
      (count, m) => count + ((m.updateCount !== 0) ? 1 : 0), 0);
  }

  toggle(): void {
    this.expanded = !this.expanded;
  }

  expandAll(): void {
    this.expanded = true;
    this.children.forEach(c => c.expandAll());
  }

  collapseAll(): void {
    this.expanded = false;
    this.children.forEach(c => c.collapseAll());
  }

}


/**
 *
 */
export class GenericSparkPlanNode extends SparkPlanNode {
  constructor(builder: SparkPlanNodeBuilder, parent: SparkPlanNode|undefined,
    info: SparkPlanInfo
  ) {
    super(builder, parent, info);
  }

}

File: spark-event-gateway-srv/src/main/java/fr/an/spark/gateway/eventTrackers/sql/SparkPlanNode.java
package fr.an.spark.gateway.eventTrackers.sql;

import fr.an.spark.gateway.eventTrackers.sql.metrics.MetricValueHolder;
import fr.an.spark.gateway.eventlog.model.SparkPlanInfo;
import lombok.Getter;

import java.util.*;
import java.util.stream.Collectors;

/**
 *
 */
@Getter
public abstract class SparkPlanNode {

    public final SparkPlanInfoTree tree;
    protected final int id;
    protected final SparkPlanNode parent;

    protected final String nodeName;
    protected final String simpleString;
    protected final Map<String, Object> metadata;

    protected final List<SparkPlanNode> children;
    protected final List<MetricValueHolder> metricValueHolders;

    protected boolean expanded = true;

    // -----------------------------------------------------------------------------------------------------------------

    public SparkPlanNode(
            SparkPlanNodeBuilder builder,
            SparkPlanNode parent,
            SparkPlanInfo info
    ) {
        this.tree = builder.getTree();
        this.id = builder.newId();
        this.parent = parent;
        this.nodeName = info.getNodeName();
        this.simpleString = info.getSimpleString();
        this.metadata = info.getMetadata();

        if (info.getChildren() != null) {
            this.children = info.getChildren().stream()
                    .map(x -> {
                        if ("InputAdapter".equals(x.getNodeName()) && x.getChildren() != null && x.getChildren().size() == 1) {
                            return builder.createNode(this, x.getChildren().get(0));
                        }
                        return builder.createNode(this, x);
                    })
                    .collect(Collectors.toList());
        } else {
            this.children = new ArrayList<>();
        }

        if (info.getMetrics() != null) {
            this.metricValueHolders = info.getMetrics().stream()
                    .map(x -> new MetricValueHolder(this, x, builder.getPreviousTree()))
                    .collect(Collectors.toList());
        } else {
            this.metricValueHolders = new ArrayList<>();
        }
    }

    public List<MetadataKeyValue> getMetadataKeyValues() {
        if (metadata == null) return Collections.emptyList();
        List<MetadataKeyValue> result = new ArrayList<>();
        for (Map.Entry<String, Object> entry : metadata.entrySet()) {
            result.add(new MetadataKeyValue(entry.getKey(), entry.getValue()));
        }
        return result;
    }

    public MetricValueHolder metricByName(String name) {
        return metricValueHolders.stream()
                .filter(x -> name.equals(x.getName()))
                .findFirst()
                .orElse(null);
    }

    public MetricValueHolder metricByAccumulatorId(int accumulatorId) {
        return metricValueHolders.stream()
                .filter(x -> x.getAccumulatorId() == accumulatorId)
                .findFirst()
                .orElse(null);
    }

    public int countMetricValuesToShow() {
        return (int) metricValueHolders.stream().filter(MetricValueHolder::isValueToShow).count();
    }

    public int countMetricsWithUpdates() {
        return (int) metricValueHolders.stream().filter(m -> m.getUpdateCount() != 0).count();
    }

    public void toggle() {
        expanded = !expanded;
    }

    public void expandAll() {
        expanded = true;
        children.forEach(SparkPlanNode::expandAll);
    }

    public void collapseAll() {
        expanded = false;
        children.forEach(SparkPlanNode::collapseAll);
    }
}
